libraries/chain/fork_database.cpp

```diff
@@ -381,9 +433,15 @@ namespace eosio { namespace chain {
     */
    pair< branch_type, branch_type >  fork_database::fetch_branch_from( const block_id_type& first,
                                                                        const block_id_type& second )const {
+      std::shared_lock g( my->mtx );
+      return my->fetch_branch_from_impl( first, second );
+   }
+
+   pair< branch_type, branch_type >  fork_database_impl::fetch_branch_from_impl( const block_id_type& first,
+                                                                                 const block_id_type& second )const {
       pair<branch_type,branch_type> result;
-      auto first_branch = (first == my->root->id) ? my->root : get_block(first);
-      auto second_branch = (second == my->root->id) ? my->root : get_block(second);
+      auto first_branch = (first == root->id) ? root : get_block_impl(first);
+      auto second_branch = (second == root->id) ? root : get_block_impl(second);
```

Comment:
Because `root` is not a complete block.

Reply:
Hum, I don't quite understand the problem. I was suggesting adding the first two lines in the following function:

```
   block_state_ptr   fork_database_impl::get_block_impl(const block_id_type& id)const {
      if (id == root->id)
         return root;
      auto itr = index.find( id );
      if( itr != index.end() )
         return *itr;
      return block_state_ptr();
   }
```

Reply:
We can't do that because `root` is not a complete block. See comment on `root` declaration and see `reset_impl`.

Reply:
But that is the same `root` that is assigned to `first_branch`, isn't it? Unless I am missing something, we are accessing the same member of the same class.

Reply:
Yes, but `get_block` is used externally as well. We could create a `get_block_or_root_impl` and use it for these cases, but not sure it is worth it.

Reply:
OK, I think I get it, there are some call locations where it would be OK to do this, and some other locations where it wouldn't. Maybe we should have two different `get_block_impl`? Still not 100% clear to me but that's OK.